sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/MetadataWhitespacePolicy.java

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.common.policy;
+
+import com.azure.core.http.HttpPipelineCallContext;
+import com.azure.core.http.HttpPipelineNextPolicy;
+import com.azure.core.http.HttpResponse;
+import com.azure.core.http.policy.HttpPipelinePolicy;
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.storage.common.implementation.Constants;
+import reactor.core.publisher.Mono;
+
+import java.util.Locale;
+
+/**
+ * This is a request policy in an {@link com.azure.core.http.HttpPipeline} to validate that metadata does not contain
+ * leading or trailing whitespace characters.
+ * We do this since the service trims whitespace for the string to sign, but the client does not, resulting in an auth
+ * failure.
+ */
+public class MetadataWhitespacePolicy implements HttpPipelinePolicy {
+
+    private final ClientLogger logger = new ClientLogger(MetadataWhitespacePolicy.class);
+
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        context.getHttpRequest().getHeaders().stream()
+            .filter(header -> header.getName().toLowerCase(Locale.ROOT).startsWith(Constants.HeaderConstants.X_MS_META))
+            .forEach(header -> {
+                String name = header.getName().substring(Constants.HeaderConstants.X_MS_META.length());
+                boolean foundWhitespace = Character.isWhitespace(name.charAt(0))
+                    || Character.isWhitespace(name.charAt(name.length() - 1));
+                for (String value: header.getValues()) {
+                    foundWhitespace |= Character.isWhitespace(value.charAt(0))
+                        || Character.isWhitespace(value.charAt(value.length() - 1));
+                }
+                if (foundWhitespace) {
+                    throw logger.logExceptionAsError(new IllegalArgumentException("Metadata keys and values can "
```

Comment:
We should return `Mono.error` here instead of throwing the error as it'll force asynchronous clients to have an additional location where it has to try/catch